src/ray/core_worker/transport/direct_actor_task_submitter.cc

```diff
@@ -497,30 +498,126 @@ void CoreWorkerDirectActorTaskSubmitter::PushActorTask(ClientQueue &queue,
   queue.rpc_client->PushActorTask(std::move(request), skip_queue, wrapped_callback);
 }
 
-void CoreWorkerDirectActorTaskSubmitter::HandlePushTaskReply(
+CoreWorkerDirectActorTaskSubmitter::PushTaskReplyAction
+CoreWorkerDirectActorTaskSubmitter::RetryOrFindAction(
     const Status &status,
     const rpc::PushTaskReply &reply,
     const rpc::Address &addr,
     const TaskSpecification &task_spec) {
+  // 1. task_skipped
+  // 1. completed (status = OK, ! reply.is_retryable_error())
+  // 1. scheduling cancelled
+  // 1. actor died (status = not OK), can retry
+  // 1. actor died (status = not OK), can't retry (may need some grace time)
+  // 1. is_retryable_error, can retry
+  // 1. is_retryable_error, can't retry
+
   const auto task_id = task_spec.TaskId();
   const auto actor_id = task_spec.ActorId();
-  const auto actor_counter = task_spec.ActorCounter();
   const auto task_skipped = task_spec.GetMessage().skip_execution();
-  /// Whether or not we will retry this actor task.
-  auto will_retry = false;
 
   if (task_skipped) {
     // NOTE(simon):Increment the task counter regardless of the status because the
     // reply for a previously completed task. We are not calling CompletePendingTask
     // because the tasks are pushed directly to the actor, not placed on any queues
     // in task_finisher_.
-  } else if (status.ok()) {
+    return PushTaskReplyAction::DO_NOTHING;
+  }
```

Comment:
since tests seem to fail badly, let's probably just revert back to original logic.. we should probably rewrite after writing unit tests

Reply:
reverted